subsys/secure_storage/Kconfig.its_transform

```diff
@@ -0,0 +1,105 @@
+# Copyright (c) 2024
+# SPDX-License-Identifier: Apache-2.0
+
+choice SECURE_STORAGE_ITS_TRANSFORM_IMPLEMENTATION
+	prompt "ITS transform module implementation"
+
+config SECURE_STORAGE_ITS_TRANSFORM_IMPLEMENTATION_AEAD
+	bool "ITS transform module implementation using AEAD to protect the data"
+
+config SECURE_STORAGE_ITS_TRANSFORM_IMPLEMENTATION_CUSTOM
+	bool "Custom ITS transform module implementation"
+	help
+	  Implement the functions declared in <zephyr/secure_storage/its/transform.h>
+	  and set CONFIG_SECURE_STORAGE_ITS_TRANSFORM_OUTPUT_OVERHEAD appropriately.
+
+endchoice # SECURE_STORAGE_ITS_TRANSFORM_IMPLEMENTATION
+
+config SECURE_STORAGE_ITS_TRANSFORM_OUTPUT_OVERHEAD
+	int "Overhead, in bytes, associated with the transformation of an entry's data for storage"
+	range 0 1000
+	# nonce (12) + authentication tag (16)
+	default 28 if SECURE_STORAGE_ITS_TRANSFORM_IMPLEMENTATION_AEAD
+	default -1
+	help
+	  This indicates how many more bytes an ITS entry's data will be once it
+	  has been processed by the secure_storage_its_transform_to_store() function.
+
+if SECURE_STORAGE_ITS_TRANSFORM_IMPLEMENTATION_AEAD
+
+choice SECURE_STORAGE_ITS_TRANSFORM_AEAD_SCHEME_PROVIDER
+	prompt "AEAD ITS transform module scheme"
+	help
+	  The AEAD scheme used to encrypt and authenticate the data.
+
+config SECURE_STORAGE_ITS_TRANSFORM_AEAD_SCHEME_PROVIDER_DEFAULT
+	prompt "Default AEAD scheme (AES-GCM)"
+	select PSA_WANT_KEY_TYPE_AES
+	select PSA_WANT_ALG_GCM
+
+config SECURE_STORAGE_ITS_TRANSFORM_AEAD_SCHEME_PROVIDER_CUSTOM
+	bool "Custom AEAD scheme"
+	help
+	  Implement the secure_storage_its_transform_aead_get_scheme() function
+	  and set CONFIG_SECURE_STORAGE_ITS_TRANSFORM_AEAD_NONCE_SIZE appropriately.
+
+endchoice # SECURE_STORAGE_ITS_TRANSFORM_AEAD_SCHEME_PROVIDER
+
+choice SECURE_STORAGE_ITS_TRANSFORM_AEAD_KEY_PROVIDER
+	prompt "AEAD ITS transform module encryption key provider"
+
+config SECURE_STORAGE_ITS_TRANSFORM_AEAD_KEY_PROVIDER_DEFAULT
+	bool "Default key provider"
+	select PSA_WANT_ALG_SHA_256
+	imply HWINFO
+	help
+	  The default key provider returns the SHA-256 hash of the UID of the ITS entry being
+	  processed as the key. In addition to the UID, it hashes the following, if available:
+	  - the device EUI64 as returned by hwinfo_get_device_eui64(), as first choice;
+	  - the device ID as returned by hwinfo_get_device_uuid(), as second choice.
+	  If no device ID is available or if the device ID is easily readable, the security of
+	  the key provided is low because the UIDs are easily guessable.
```

Comment:
From what I understand, the uid is stored in plaintext(as additional data) so the hash could be calculated directly and `easily guessable` is not really true for the case without device id. We might modify that a bit, but haven't given it much thought yet.
Also, could we drop the implementation without hardware id, in case I'd open a PR to add an implementation for native posix, or would that be needed elsewhere?

Reply:
> From what I understand, the uid is stored in plaintext(as additional data)

No, it's not. It's used along with `create_flags` as additional data to authenticate, but the additional data is not part of the resulting ciphertext. You can see that `create_flags` is manually added to the `stored_entry` struct next to the ciphertext because we need to store it, but the UID is not found anywhere in the `stored_entry`.

(The reason why `create_flags` is not simply encrypted with the entry's data is that it would require using one more buffer, in which the data and `create_flags` would be put together, that would then be passed to PSA Crypto for the AEAD operation. It's avoided to be less hard on memory as the stack is already under heavy usage in there, and `create_flags` does not need to be kept secret. However we do want to authenticate both `create_flags` and the entry UID, thus they make up the additional data.)

Now, the entry UIDs actually do (may) get stored in plaintext by the store module. It needs to use them to be able to retrieve a particular entry. The current settings store module implementation directly uses the entry UIDs as setting names, so indeed `the key(hash) could be calculated directly`.

> and `easily guessable` is not really true for the case without device id. We might modify that a bit, but haven't given it much thought yet.

You're right. This description was left from an earlier version where I had encryption of the setting names (you can see it [here](https://github.com/tomi-font/zephyr/blob/e973b7714237d45d604107327cc7569a07ba0a70/subsys/secure_storage/src/its/storing_settings.c#L15-L64)). I used AES-ECB to encrypt the (single-block) entry UIDs used in the setting names with the same key used for encrypting the data when this insecure key provider was in use. That way, the UIDs were not actually stored in clear anywhere.
But because this was still not secure (just some more obfuscated) as the UIDs are easily guessable, I ended up removing it thinking it would be frowned upon. It could easily be brought back though.

> Also, could we drop the implementation without hardware id, in case I'd open a PR to add an implementation for native posix, or would that be needed elsewhere?

What would the implementation be like? And it would be only for simulation platforms?
The goal is to standardize on the PSA Secure Storage API, so it should be usable on all platforms, even if that means having only functional support (and no security) on certain ones. Even though the HW info API has drivers for a bunch of platforms, I don't think it's all and as such it's better to have a fallback solution for compatibility.

Reply:
> No, it's not. ... create_flags is manually added to the stored_entry.
Indeed, somehow I overlooked that and thought that both were added to the stored entry. I shouldn't have rushed the review submission in the morning. Sorry.

> , the entry UIDs actually do (may) get stored in plaintext by the store module. 
...
I used AES-ECB to encrypt the (single-block) entry UIDs used in the setting names with the same key used for encrypting the data when this insecure key provider was in use.

A few thoughts:
Looking at the Deployment models of arm's Security Risk Assessment of the API I think the current implementation goes into the direction of `DM.PROTECTED` but with no isolation. In case the the key is not derivable from the content of the storage medium, we could also support a deployment model similar to the mentioned `DM.EXPOSED`, but again without any isolation.
Additionally, in case a driver kernel would be implemented and user space enabled we would additionally be able to provide some isolation for user threads.

So believe we would need/(would be great) to define define the security objectives, requirements and threats for a few deployment scenarios we have in Zephyr to decide what protection mechanisms are needed and bring a benefit for each.

But that would probably take while. So I am curious about your timeline. I guess your priority is be to merge this soon with a basic implementation. To be able to convert the in-tree code and improve and provide additional implementations later on.

Reply:
> Looking at the Deployment models of arm's Security Risk Assessment of the API I think the current implementation goes into the direction of DM.PROTECTED but with no isolation. In case the the key is not derivable from the content of the storage medium, we could also support a deployment model similar to the mentioned DM.EXPOSED, but again without any isolation.

It's a bit hard as the end result will vary depending on what the end user chooses to do, but I'd say that right now the base implementation in this PR is more similar to `DM.EXPOSED`, without isolation. It's purely security at rest for the data.
The case where the key is derivable from the contents of the storage medium is really just for compatibility, and end users are urged to at least implement a secure key provider.

> Additionally, in case a driver kernel would be implemented and user space enabled we would additionally be able to provide some isolation for user threads.

Kernel and user space separation has already been mentioned, and it could be something to look into if we want to try to provide some sort of isolation.

> So believe we would need/(would be great) to define define the security objectives, requirements and threats for a few deployment scenarios we have in Zephyr to decide what protection mechanisms are needed and bring a benefit for each.

It could make sense. Especially if/when different implementations are made available (e.g. one where the storage medium is protected by the HW).

> But that would probably take while. So I am curious about your timeline. I guess your priority is be to merge this soon with a basic implementation. To be able to convert the in-tree code and improve and provide additional implementations later on?

Yeah, I think it would be good to get an initial, relatively basic implementation of the PSA Secure Storage API merged first, and to build on top of it. Having the API available in Zephyr will help other areas, one being the adoption of the PSA Crypto API.